### ISSUE TITLE
Need Conda package for pandaSDMX

### ISSUE THREAD

**Issue by alcibiade** [dr-leo/pandaSDMX#43]
Conda is the de-facto packaging and distribution of dependencies for Jupyter Notebook components. PIP is not installed as a default on the datascience workbook images: https://github.com/jupyter/docker-stacks/tree/master/datascience-notebook

It is still possible to work around this by forcing PIP install and install pandaSDMX:

```
FROM jupyter/datascience-notebook

RUN conda install -y pip
RUN pip install pandasdmx
```

We should certainly have an up-to-date package in Conda as data science notebooks are probably a major platform for users of pandaSDMX (as myself).

I found this documentation on distributing an existing PIP-ready project to conda:
http://conda.pydata.org/docs/build_tutorials/pkgs.html

@dr-leo what's your point on this ?

**Comment by dr-leo** [comment]
Hi,

thanks. I'd very much like to make pandSDMX conda-installable. And the
guide seems straightforward. I'll look into this once I find the time.

Leo

**Comment by dr-leo** [comment]
Hi,


I've tried this vainly. Alas, I work on Windows where conda skeleton
needs a MinGW build env. Well, I do have Docker and that sort of stuff.
But as a seasoned programmer it would take me forever to get started
with Linux on Virtual Box etc.


So if you are in the mood to give it a try building a conda pkg and
uploading it to anaconda.org, I would applaud.

Leo

Am 09.01.2017 um 11:58 schrieb Alcibiade:

**Comment by alcibiade** [comment]
I'm using Linux as my main OS so I guess that it'll be easier from here.

I'm going to give it a try and let you know about it.

**Comment by alcibiade** [comment]
I have built a Linux-64 Conda package of pandaSDMX 0.6 which is available here:
https://anaconda.org/alcibiade/pandasdmx

It can be installed like this:
```
conda install -c alcibiade pandasdmx=0.6
```

I don't know yet if it can be promoted to the main repository (instead of my user's namespace), I'll dig further.

I didn't build it for all the architectures, but it looks like another guy did the job for the version 0.3.1:
https://anaconda.org/delgadom/pandasdmx

I'm going to try to contact him to see if he used some kind of specific tool to automate multi-architecture builds.

**Comment by dr-leo** [comment]
Hi,

the build documentation you retrieved contains a section on converting
a pkg to other platforms. This should make Automation easy:

Leo

Optional - Convert conda package for other platforms

Now that you have built a package for your current platform with
conda-build, you can convert it for use on other platforms with the
conda convert command and a platform specifier from the list
{osx-64,linux-32,linux-64,win-32,win-64,all}. In the output directory,
one folder will be created for each of the one or more platforms you
chose, and each folder will contain a .tar.bz2 package file for that
platform.

Windows users:


conda convert -f --platform all
C:\Users\jsmith\Miniconda\conda-bld\win-64\pyinstrument-0.13.1-py27_0.tar.bz2
-o outputdir\


Linux and OS X users:


conda convert --platform all
/home/jsmith/miniconda/conda-bld/linux-64/pyinstrument-0.13.1-py27_0.tar.bz2
-o outputdir/


NOTE: Change your path and filename to the exact path and filename you
saved in Step 2.

**Comment by alcibiade** [comment]
Rookie mistake, I didn't read the doc till the end :-)

All architectures are now available.

**Comment by dr-leo** [comment]
Hi,


thanks. That's great! I've done a successful install with py35.


However, there is no package for py27; I tried this unsuccessfully.

So assuming you created packages for architectures win32, win64, linux32
and linux64 (plus macos?), and assuming further that py27, 34 to 36
should be supported, there should be at least 16, if not 20 or more
packages required for proper conda support. This can clearly be
automated easily as described in the guide. Do you think you could give
it a try? - pandasdmx supports all these python versions with the same
code base.
 
Once this is done, I will update the pandasdmx docs. There should also
be a quick install heading on the front page. Finally, we might approach
continuum.io for inclusion in the top level namespace.
Thanks again and keep up your great work!

Leo


 
 
Am 27.01.2017 um 21:14 schrieb Alcibiade:

**Comment by dr-leo** [comment]
Re-opening as Python 27 and 34-36 should be supported.

**Comment by dr-leo** [comment]
Hi,


while I could install v0.6 from your anaconda channel, I cannot import
it. The pandasdmx subdir is not in site-packages. I cannot dig deeper
now due to time constraints. But "conda uninstall pandasdmx" reports
that it is installed. Don't know what's going on here.


Leo 





Am 27.01.2017 um 21:14 schrieb Alcibiade:

**Comment by alcibiade** [comment]
I'm going to review this ASAP. I only tested the linux-64 architecure which works well, but maybe there was an issue with the cross-platform generation.

I'll see to build Python 2.7 packages too also.

**Comment by alcibiade** [comment]
After checking, the previous versions of the packages were broken, except the original one linux-64 because I had a failure before generating the other architectures. I think the packages were generated without the pandasdmx package folder.

Now this is fixed, and we have packages for most architectures and both Python 2.7 and Python 3.5

I guess you could help by re-installing the windows library again. It should work with the latest version.

**Comment by dr-leo** [comment]
Hi,


thanks. Now, packages for py27 and 35 are there. But neither installs.
Here's the traceback - maybe it is a known pitfall in package creation.
I consider a conda bug as unlikely.


Thanks for your perseverance!


Leo


Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment
C:\Users\stefan\Anaconda3\envs\py2:


The following NEW packages will be INSTALLED:

    pandasdmx: 0.6-py27_0 alcibiade

Proceed ([y]/n)? y

pandasdmx-0.6- 100% |###############################| Time: 0:00:00
129.93 kB/s
An unexpected error has occurred.
Please consider posting the following information to the
conda GitHub issue tracker at:

    https://github.com/conda/conda/issues



Current conda install:

               platform : win-32
          conda version : 4.3.8
       conda is private : False
      conda-env version : 4.3.8
    conda-build version : 2.1.2
         python version : 3.4.5.final.0
       requests version : 2.12.4
       root environment : C:\Users\stefan\Anaconda3  (writable)
    default environment : C:\Users\stefan\Anaconda3\envs\py2
       envs directories : C:\Users\stefan\Anaconda3\envs
          package cache : C:\Users\stefan\Anaconda3\pkgs
           channel URLs : https://conda.anaconda.org/alcibiade/win-32
                          https://conda.anaconda.org/alcibiade/noarch
                          https://repo.continuum.io/pkgs/free/win-32
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/r/win-32
                          https://repo.continuum.io/pkgs/r/noarch
                          https://repo.continuum.io/pkgs/pro/win-32
                          https://repo.continuum.io/pkgs/pro/noarch
                          https://repo.continuum.io/pkgs/msys2/win-32
                          https://repo.continuum.io/pkgs/msys2/noarch
            config file : None
           offline mode : False
             user-agent : conda/4.3.8 requests/2.12.4 CPython/3.4.5
Windows/7 Wi
ndows/6.1.7601



`$ C:\Users\stefan\Anaconda3\Scripts\conda-script.py install -c
alcibiade pandas
dmx`




    Traceback (most recent call last):
      File
"C:\Users\stefan\Anaconda3\lib\site-packages\conda\exceptions.py", li
ne 617, in conda_exception_handler
        return_value = func(*args, **kwargs)
      File
"C:\Users\stefan\Anaconda3\lib\site-packages\conda\cli\main.py", line
 137, in _main
        exit_code = args.func(args, p)
      File
"C:\Users\stefan\Anaconda3\lib\site-packages\conda\cli\main_install.p
y", line 80, in execute
        install(args, parser, 'install')
      File
"C:\Users\stefan\Anaconda3\lib\site-packages\conda\cli\install.py", l
ine 347, in install
        execute_actions(actions, index, verbose=not context.quiet)
      File "C:\Users\stefan\Anaconda3\lib\site-packages\conda\plan.py",
line 837
, in execute_actions
        execute_instructions(plan, index, verbose)
      File
"C:\Users\stefan\Anaconda3\lib\site-packages\conda\instructions.py",
line 258, in execute_instructions
        cmd(state, arg)
      File
"C:\Users\stefan\Anaconda3\lib\site-packages\conda\instructions.py",
line 119, in UNLINKLINKTRANSACTION_CMD
        txn.execute()
      File
"C:\Users\stefan\Anaconda3\lib\site-packages\conda\core\link.py", lin
e 255, in execute
        self.verify()
      File
"C:\Users\stefan\Anaconda3\lib\site-packages\conda\core\link.py", lin
e 243, in verify
        self.num_unlink_pkgs),
      File
"C:\Users\stefan\Anaconda3\lib\site-packages\conda\core\link.py", lin
e 240, in <genexpr>
        exceptions = tuple(exc for exc in concatv(
      File
"C:\Users\stefan\Anaconda3\lib\site-packages\conda\core\link.py", lin
e 188, in _verify_individual_level
        log.debug(format_exc())
      File "C:\Users\stefan\Anaconda3\lib\traceback.py", line 256, in
format_exc

        return "".join(format_exception(*sys.exc_info(), limit=limit,
chain=chai
n))
      File "C:\Users\stefan\Anaconda3\lib\traceback.py", line 181, in
format_exc
eption
        return list(_format_exception_iter(etype, value, tb, limit, chain))
      File "C:\Users\stefan\Anaconda3\lib\traceback.py", line 146, in
_format_ex
ception_iter
        for value, tb in values:
      File "C:\Users\stefan\Anaconda3\lib\traceback.py", line 125, in
_iter_chai
n
        context = exc.__context__
    AttributeError: 'NoneType' object has no attribute '__context__'


(py2) C:\users\stefan\documents\dev\pysdmx>


Am 30.01.2017 um 21:47 schrieb Alcibiade:

**Comment by alcibiade** [comment]
I don't really understand your error... Could it be related to filesystem error as we were updating the same release file (this could lock files on windows and prevent overwrite) ? 

I got access to a windows system and installed successfully on 2 different configurations:

Miniconda 32 bits / Python2.7:

![capture](https://cloud.githubusercontent.com/assets/3389277/22598487/ef756d18-ea33-11e6-9215-05df8413f9df.PNG)

The example code failed but I don't think it is related to the conda install.

Miniconda 64 bits / Python3.5:

![capture2](https://cloud.githubusercontent.com/assets/3389277/22598511/f4fd1b00-ea33-11e6-9200-81a950ded68b.PNG)

**Comment by dr-leo** [comment]
I've seen this type of error on Windows before, but the files were always properly written. Be it as it may: I've submitted an issue for this:
https://github.com/conda/conda/issues/4519

I work on Win7, 64bis. 

Given that you've tried it on Windows successfully, I would not mind updating the pandasdmx docs. Have you built packages for py27, 34, 35 and 36? I would consider this a must-have.

On the error you may have encountered on Py27: There was a 2to3 issue in remote.py. I fixed it and will do a bug fix release v0.6.1 in the hope you do not bother rebuilding the conda packages.

**Comment by dr-leo** [comment]
v0.6.1 is now on pypi. I've update the docs to reference the conda
install option through your channel. Once you confirm that there are
conda packages for all relevant py versions, we can approach
Continuum.io to ask re. global namespace.


Am 03.02.2017 um 17:16 schrieb Alcibiade:

**Comment by alcibiade** [comment]
Done for all python versions and architectures: https://anaconda.org/alcibiade/pandasdmx/files

For the next release I will prepare a script.

As owner of the project I guess it's best that you contact Continuum directly for the namespace.

**Comment by dr-leo** [comment]
Wow! That's impressive! - As you may have seen, I've updated the docs.
- I will also Approach Continuum re. Namespace.

I hope the next release will be a few months down the road - Hopefully
there isn't any severe bug that I've overlooked.

Keep up your great work!

Cheers,

Leo

**Comment by dr-leo** [comment]
Hi Alcibiade,

I hope this finds you well.

Just to follow up:

On making pandaSDMX a top Level conda package: After some research I
have given up. Don't know whom to contact at Continuum.io.

I hope to do the 0.7 release in the next few weeks. It will Support 4
additional data Providers which is a big step. Besides, the API has
been considerably improved in Terms of bettr key Validation,
connecting to custom agencies etc. Finally, a lot of Bugs have been
fixed, notably in the json Reader.

Would you be ready to build the conda packages as last time?

Best regards

Leo

**Comment by alcibiade** [comment]
Hi there,

Yes, no problem. Just let me know when the new release is ready and I'll do
the builds.
I just need to find the build script I used the last time :-)

Regards,

Yannick

--
Yannick Kirschhoffer

e-Mail: alcibiade@alcibiade.org
Twitter: @Kirschhoffer <https://twitter.com/kirschhoffer>
LinkedIn: linkedin.com/in/alcibiade

2017-05-12 10:55 GMT+02:00 dr-leo <notifications@github.com>:

**Comment by dr-leo** [comment]
Having googled a bit I realize that ultimately conda packages of
pandaSDMX should live on conda-forge:

http://conda-forge.github.io/docs/


This is all new to me and I doubt I'll ever find the time to dig deep
enough to make this fly.


My best guess is that the upcoming version could be in your channel. But
future versions should at some point migrate to conda-forge.


If you are interested in undertaking this effort, please let me know.
PR's are welcome. The infrastructure looks fascinating and there's a lot
to learn about CI and other stuff.


What do you think? 


Leo

Am 12.05.2017 um 18:21 schrieb Alcibiade:

**Comment by dr-leo** [comment]
Hi,

I have just released v0.7.0 on pypi. I would be thrilled if you could update your conda packages.

Thanks and best regards

Leo

**Comment by alcibiade** [comment]
Hi,

I have updated the builds.
I had a few issues:
- The dependency on jsonpath-rw had to be built explicitly. I guess it is
because it is not available as a core conda package
- I can't manage to make the Python 2.7 build work consistently for now. I
only distributed version 3.x for now

I will try to find a solution for the 2.7 build asap.

For now builds are available on https://anaconda.org/alcibiade/
I have put the current build script in a fork of the project on
https://github.com/alcibiade/pandaSDMX/tree/master/scripts

This way it will be faster to build the next updates :-)

Regards,

--
Yannick Kirschhoffer

e-Mail: alcibiade@alcibiade.org
GitHub: https://github.com/alcibiade
LinkedIn: linkedin.com/in/alcibiade

2017-06-10 13:07 GMT+02:00 dr-leo <notifications@github.com>:
